plc/int6kmod.1
.TH int6kmod 1 "November 2013" "open-plc-utils-0.0.3" "Qualcomm Atheros Open Powerline Toolkit"

.SH NAME
int6kmod - Qualcomm Atheros Powerline Device Module Manager

.SH SYNOPSIS
.BR int6kmod
.RI [ options ] 
.RI [ device ] 
.RI [ device ] 
[...]

.SH DESCRIPTION
Read or write Atheros custom powerline modules using VS_MODULE_OPERATION messages.
Modules may be read from a device and written to a file or read from a file and written to a device.
Modules may have any of several formats and purposes.
Where the firmware permits, this program can now read modules in excess of 1400 bytes and can write multiple modules that exceed 1400 bytes in one operation.

.PP
This program is part of the Qualcomm Atheros Powerline Toolkit.
See the \fBplc\fR man page for an overview and installation instructions.

.SH OPTIONS

.TP
.RB - d
Read the specified module from NVRAM and write to standard output in hexadecimal dump format.
The module is identified by module and sub-module using options \fB-t\fR and \fB-s\fR.
Runtime firmware will report an error if the module is not available.
The session identifier is not required for a module read operation and is ignored.
The largest module that can be read is 1400 bytes.
This option and option \fB-m\fR are mutually exclusive.

.TP
.RB - e
Redirects stderr messages to stdout.
By convention status and error messages are printed on stderr while primary program output is printed on stdout.
This option prints all output on stdout in cases where this is desired.

.TP
-\fBi \fIinterface\fR
Select the host Ethernet interface.
All requests are sent via this host interface and only reponses received via this host interface are recognized.
The default interface is \fBeth1\fR because most people use \fBeth0\fR as their principle network connection; however, if environment string "PLC" is defined then it takes precedence over the default interface.
This option then takes precedence over either default.

.TP
-\fBm \fIfilename\fR
Read the specified module from NVRAM and write it to the named file.
The module is identified by module and sub-module using options \fB-t\fR and \fB-s\fR.
Runtime firmware will report an error if the module is not available.
The session identifier is not required for a module read operation and is ignored.
The largest module that can be read is 1400 bytes.
This option and option \fB-d\fR are mutually exclusive.

.TP
-\fBM \fIfilename\fR
Read the named module file and download it to the named device and commit NVRAM.
The module is assigned to a module and sub-module identifier using options \fB-t\fR and \fB-s\fR.
Runtime firmware will reject modules having invalid module and sub-mocule identifiers.
Module files must be a multiple of four bytes or an error will occur.
As implemented, int6kmod will reject files exceeding 1400 bytes but this restriction will be removed in later versions.

.TP
.RB - q
Suppresses status messages on stderr.

.TP
-\fBs \fIsub-module-id\fR
The sub-module identifier expressed as a 16-bit hexadecimal value with optional \fB0x\fR prefix.
This identifier indicates a specific module with the module class specified using option \fB-t\fR.
Any 16-bit value is permitted but only certain values are accepted by device firmware depending on firmware type and revision.
The default value is \fB0x0000\fR.

.TP
-\fBS \fIsession-id\fR
The session identifier expressed as a 32-bit hexadecimal value with optional \fB0x\fR prefix.
The session identifier value is arbitrary and so any 32-bit value is permitted but 0x00000000 and 0xFFFFFFFF are illegal.
The \fB0x\fR prefix is optional for this argument.
The default \fIsession-id\fR is 0x78563412.

.TP
-\fBt \fImodule-id\fR
The module identifier expressed as a 16-bit hexadecimal value with optional \fB0x\fR prefix.
This identifier indicates the module type or class.
Option \fB-s\fR indicates the specific member of a module class.
Any 16-bit value is permitted but only certain values will be accepted by device firmware.
The default value is \fB0x0000\fR.

.TP
-\fBv\fR
Print additional information on stdout.
In particular, this option dumps incoming and outgoing packets which can be saved as text files for reference.

.TP
.RB - w
Write settings to the device instead of reading settings from the device.
If omitted then settings are read from the device and displayed.
If present, the user should specify all settings to avoid writing default values to the device.

.TP
.RB - ? ,-- help
Print program help summary on stdout.
This option takes precedence over other options on the command line.

.TP
.RB - ? ,-- version
Print program version information on stdout.
This option takes precedence over other options on the command line.
Use this option when sending screen dumps to Atheros Technical Support so that they know exactly which version of the Linux Toolkit you are using.

.SH ARGUMENTS

.TP
.IR device
The MAC address of some powerline device.
More than one address may be specified on the command line.
If more than one address is specified then operations are performed on each device in turn.
The default address is \fBlocal\fR.
See \fBDEVICES\fR for information.

.SH DEVICES
Powerline devices use Ethernet Media Access Control (MAC) addresses.
A MAC address is a 48-bit value entered as 12 hexadecimal digits in upper, lower or mixed character case.
Octets may be separated with colons for clarity.
For example, "00b052000001", "00:b0:52:00:00:01" and "00b052:000001" are valid and equivalent.

.PP
The following MAC addresses are special and may be entered by name instead of number.

.TP
.BR all
Same as "broadcast".

.TP
.BR broadcast
A synonym for the Ethernet broadcast address, \fBFF:FF:FF:FF:FF:FF\fR.
All devices, whether local, remote or foreign recognize messages sent to this address.
A remote device is any device at the far end of a powerline connection.
A foreign device is any device not manufactured by Atheros.

.TP
.BR local
A synonym for the Qualcomm Atheros vendor specific Local Management Address (LMA), \fB00:B0:52:00:00:01\fR.
All local Atheros devices recognize this address but remote and foreign devices do not.
A remote device is any device at the far end of a powerline connection.
A foreign device is any device not manufactured by Atheros.

.SH REFERENCES
See the Qualcomm Atheros HomePlug AV Firmware Technical Reference Manual for more information.

.SH DISCLAIMER
Atheros HomePlug AV Vendor Specific Management Message structure and content is proprietary to Qualcomm Atheros, Ocala FL USA.
Consequently, public information is not available.
Qualcomm Atheros reserves the right to modify message structure or content in future firmware releases without any obligation to notify or compensate users of this program.

.SH EXAMPLES
The following command reads module file \fBabc.bin\fR and writes it to the local device.
The module identifier \fB0x1000\fR indicates that it is an \fBMDIO Init\fR module.
The sub-module identifer defaults to \fB0x0000\fR because option \fB-s\fR is omitted.
The session identifier defaults to \fB0x78563412\fR because option \fB-S\fR is omitted.
The operation takes place in three steps.
The last step may take 2 or 3 seconds to complete.

.PP
   # int6kmod -M abc.bin -t 0x1000 
   eth1 00:B0:52:00:00:01 Request Session
   eth1 00:B0:52:BE:EF:03 Request Granted
   eth1 00:B0:52:00:00:01 Write abc.bin
   eth1 00:B0:52:BE:EF:03 Written
   eth1 00:B0:52:00:00:01 Commit Modules
   eth1 00:B0:52:BE:EF:03 Committed

.PP
The next example does the same thing but writes module file \fBdef.bin\fR to remote device \fB00:B0:52:BA:BE:02\fR.
The module identifier \fB0x4000\fR indicates that it is an \fBAdvanced Power Management uC\fR module.
The session identifier used is \fB0xA5A5A5A5\fR.
The sub-module identifer defaults to \fB0x0000\fR because option \fB-s\fR is omitted.

.PP
   # int6kmod -M def.bin -t 0x4000 -S A5A5A5A5 00:B0:52:BA:BE:02
   eth1 00:B0:52:00:00:01 Request Session
   eth1 00:B0:52:BE:EF:03 Request Granted
   eth1 00:B0:52:00:00:01 Write abc.bin
   eth1 00:B0:52:BE:EF:03 Written
   eth1 00:B0:52:00:00:01 Commit Modules
   eth1 00:B0:52:BE:EF:03 Committed

.PP
The next example reads back the module written in the first example and saves it in file \fBzyx.bin\fR.

.PP
   # int6kmod -m xyz.bin -t 0x1000
   eth1 00:B0:52:00:00:01 Read zyx.bin
   eth1 00:B0:52:BE:EF:03 Read

.SH SEE ALSO
.BR plc ( 1 ),
.BR int6kmdio ( 1 ),
.BR int6kmdio2 ( 1 ),
.BR mdioblock ( 1 ),
.BR mdioblock2 ( 1 ),
.BR mdiodump ( 1 ),
.BR mdiogen ( 1 )

.SH CREDITS
 Charles Maier



CHANGES
open-plc-utils-0.0.6
---------------

1. Programs mac2pw and mac2pwd implement a new method of generating device passwords. The old method is supported but is not the default. 

open-plc-utils-0.0.5
---------------

open-plc-utils-0.0.4
---------------

open-plc-utils-0.0.3
---------------

1. Added header VisualStudioNET/include/inttype.h to define print specifier constant PRId64 for Windows systems.

2. Replaced conditional compiler code around print specifiers "lld" and "I64d" and replaced single specifier PRId64 as defined by the system.

3. Renamed man pages ending in .7 with .1 to reflect that fact that they will now be installed man1 instead of man7.

4. Replaced const HOMEPLUG_MTYPE with constant ETH_P_HPAV throughout.

5. Correct endianess of FIELD_ETH_TYPE in module coqos_add.c.

6. Updated chipset codes and implemented table-driven chipset detection algorithm.

7. Moved TLVNode structure definition from nvm/nvm.h to tools/tlv.h to prepare for exanded use of TLV structures in future MMEs.


open-plc-utils-0.0.2
---------------

1. Removed references to private vendor-specific MMEs

2. Removed unused archive targets in primary Makefile.

3. Added .gitignore to subdirectories where programs and other objects are compiled.

4. Added conditional __UCLIBC__ compile statement to ARCPPrint.c to account for variable argument definition error during compile; 

5. Added 'ignore' target to cascading Makefiles to re-write local .gitignore files.

6. Corrected email address for Florian Fainelli in module ARPCPrint.

7. Added new HomePlug AV SLAC Protocol MMEs definitions to homeplug.h.

8. Modified makefile 'ignore' target to use tr instead of sed because some sed implementations do not accept '\n'.



plc/int6kmod.c
/*====================================================================*
 *
 *   Copyright (c) 2013 Qualcomm Atheros, Inc.
 *
 *   All rights reserved.
 *
 *   Redistribution and use in source and binary forms, with or
 *   without modification, are permitted (subject to the limitations
 *   in the disclaimer below) provided that the following conditions
 *   are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials
 *     provided with the distribution.
 *
 *   * Neither the name of Qualcomm Atheros nor the names of
 *     its contributors may be used to endorse or promote products
 *     derived from this software without specific prior written
 *     permission.
 *
 *   NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
 *   GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE
 *   COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 *   IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 *   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
 *   OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 *   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 *   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 *   OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *--------------------------------------------------------------------*/

/*====================================================================*"
 *
 *    int6kmod.c -
 *
 *
 *   Contributor(s):
 *      Charles Maier
 *
 *--------------------------------------------------------------------*/


/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <limits.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/putoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/symbol.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../tools/error.h"
#include "../ether/channel.h"
#include "../plc/plc.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../plc/Devices.c"
#include "../plc/Confirm.c"
#include "../plc/Display.c"
#include "../plc/Request.c"
#include "../plc/Failure.c"
#include "../plc/ReadMME.c"
#include "../plc/SendMME.c"
#include "../plc/ModuleCommit.c"
#include "../plc/ModuleRead.c"
#include "../plc/ModuleDump.c"
#include "../plc/ModuleSpec.c"
#include "../plc/ModuleSession.c"
#include "../plc/ModuleWrite.c"
#endif

#ifndef MAKEFILE
#include "../tools/basespec.c"
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/hexdump.c"
#include "../tools/hexview.c"
#include "../tools/hexencode.c"
#include "../tools/hexdecode.c"
#include "../tools/hexstring.c"
#include "../tools/todigit.c"
#include "../tools/checkfilename.c"
#include "../tools/checksum32.c"
#include "../tools/fdchecksum32.c"
#include "../tools/strfbits.c"
#include "../tools/synonym.c"
#include "../tools/typename.c"
#include "../tools/error.c"
#endif

#ifndef MAKEFILE
#include "../ether/openchannel.c"
#include "../ether/closechannel.c"
#include "../ether/readpacket.c"
#include "../ether/sendpacket.c"
#include "../ether/channel.c"
#endif

#ifndef MAKEFILE
#include "../mme/MMECode.c"
#include "../mme/EthernetHeader.c"
#include "../mme/QualcommHeader.c"
#include "../mme/UnwantedMessage.c"
#endif

/*====================================================================*
 *
 *   void Manager (struct plc * plc, uint16_t ModuleID, uint16_t ModuleSubID);
 *
 *   read a custom modul from flash memory or write a suceom module
 *   to flash memory using the VS_MODULE_OPERATION message; read and
 *   save the old module before writing and commiting the new module;
 *
 *   ModuleDump () and ModuleRead () are called with argument source
 *   set to VS_MODULE_FLASH because this function operates on modules
 *   in flash memory only;
 *
 *
 *--------------------------------------------------------------------*/

static void Manager (struct plc * plc, uint16_t ModuleID, uint16_t ModuleSubID)

{
	if (_anyset (plc->flags, PLC_DUMP_MODULE))
	{
		ModuleDump (plc, PLC_MOD_OP_READ_FLASH, ModuleID, ModuleSubID);
		return;
	}
	if (_anyset (plc->flags, PLC_READ_MODULE))
	{
		ModuleRead (plc, &plc->nvm, PLC_MOD_OP_READ_FLASH, ModuleID, ModuleSubID);
	}
	if (_anyset (plc->flags, PLC_WRITE_MODULE))
	{
		struct vs_module_spec vs_module_spec;
		vs_module_spec.MODULE_ID = ModuleID;
		vs_module_spec.MODULE_SUB_ID = ModuleSubID;
		ModuleSpec (&plc->NVM, &vs_module_spec);
		ModuleSession (plc, 1, &vs_module_spec);
		ModuleWrite (plc, &plc->NVM, 0, &vs_module_spec);
		ModuleCommit (plc, (PLC_COMMIT_FORCE));
		return;
	}
	return;
}


/*====================================================================*
 *
 *   int main (int argc, char const * argv[]);
 *
 *   parse command line, populate plc structure and perform selected
 *   operations; show help summary if asked; see getoptv and putoptv
 *   to understand command line parsing and help summary display; see
 *   plc.h for the definition of struct plc;
 *
 *   the command line accepts multiple MAC addresses and the program
 *   performs the specified operations on each address, in turn; the
 *   address order is significant but the option order is not; the
 *   default address is a local broadcast that causes all devices on
 *   the local H1 interface to respond but not those at the remote
 *   end of the powerline;
 *
 *   the default address is 00:B0:52:00:00:01; omitting the address
 *   will automatically address the local device; some options will
 *   cancel themselves if this makes no sense;
 *
 *   the default interface is eth1 because most people use eth0 as
 *   their principle network connection; you can specify another
 *   interface with -i or define environment string PLC to make
 *   that the default interface and save typing;
 *
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv [])

{
	extern struct channel channel;
	static char const * optv [] =
	{
		"dei:m:M:qs:S:t:vx",
		"device [device] [...]",
		"Qualcomm Atheros Powerline Device Module Manager",
		"d\tdump module in hexadecimal format",
		"e\tredirect stderr to stdout",

#if defined (WINPCAP) || defined (LIBPCAP)

		"i n\thost interface is (n) [" LITERAL (CHANNEL_ETHNUMBER) "]",

#else

		"i s\thost interface is (s) [" LITERAL (CHANNEL_ETHDEVICE) "]",

#endif

		"m f\tread module from NVRAM to file (f)",
		"M f\twrite module file (f) to flash memory",
		"s n\tmodule sub-ID [" LITERAL (PLC_SUBMODULEID) "]",
		"S n\tsession ID [" LITERAL (PLCSESSION) "]",
		"t n\tmodule ID [" LITERAL (PLC_MODULEID) "]",
		"q\tquiet mode",
		"v\tverbose mode",
		(char const *) (0)
	};

#include "../plc/plc.c"

	struct vs_module_spec vs_module_spec;
	uint16_t ModuleID = 0;
	uint16_t ModuleSubID = 0;
	signed c;
	memset (&vs_module_spec, 0, sizeof (vs_module_spec));
	if (getenv (PLCDEVICE))
	{

#if defined (WINPCAP) || defined (LIBPCAP)

		channel.ifindex = atoi (getenv (PLCDEVICE));

#else

		channel.ifname = strdup (getenv (PLCDEVICE));

#endif

	}
	optind = 1;
	while ((c = getoptv (argc, argv, optv)) != -1)
	{
		switch (c)
		{
		case 'e':
			dup2 (STDOUT_FILENO, STDERR_FILENO);
			break;
		case 'C':
			_setbits (plc.flags, PLC_COMMIT_MODULE);
			break;
		case 'd':
			if (_anyset (plc.flags, PLC_READ_MODULE))
			{
				error (1, EINVAL, "Options -d and -m are mutually exclusive");
			}
			_setbits (plc.flags, PLC_DUMP_MODULE);
			break;
		case 'i':

#if defined (WINPCAP) || defined (LIBPCAP)

			channel.ifindex = atoi (optarg);

#else

			channel.ifname = optarg;

#endif

			break;
		case 'M':
			_setbits (plc.flags, PLC_WRITE_MODULE);
			if (!checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.NVM.file = open (plc.NVM.name = optarg, O_BINARY|O_RDONLY)) == -1)
			{
				error (1, errno, "%s", plc.NVM.name);
			}
			if (ModuleSpec (&plc.NVM, &vs_module_spec) == -1)
			{
				error (1, errno, "%s", optarg);
			}
			break;
		case 'm':
			if (_anyset (plc.flags, PLC_DUMP_MODULE))
			{
				error (1, EINVAL, "Options -d and -m are mutually exclusive");
			}
			_setbits (plc.flags, PLC_READ_MODULE);
			if (!checkfilename (optarg))
			{
				error (1, EINVAL, "%s", optarg);
			}
			if ((plc.nvm.file = open (plc.nvm.name = optarg, O_BINARY|O_CREAT|O_RDWR|O_TRUNC, FILE_FILEMODE)) == -1)
			{
				error (1, errno, "%s", plc.nvm.name);
			}

#ifndef WIN32

			chown (optarg, getuid (), getgid ());

#endif

			break;
		case 't':
			ModuleID = (uint16_t)(basespec (optarg, 16, sizeof (ModuleID)));
			break;
		case 's':
			ModuleSubID = (uint16_t)(basespec (optarg, 16, sizeof (ModuleSubID)));
			break;
		case 'S':
			plc.cookie = (uint32_t)(basespec (optarg, 16, sizeof (plc.cookie)));
			break;
		case 'q':
			_setbits (channel.flags, CHANNEL_SILENCE);
			_setbits (plc.flags, PLC_SILENCE);
			break;
		case 'v':
			_setbits (channel.flags, CHANNEL_VERBOSE);
			_setbits (plc.flags, PLC_VERBOSE);
			break;
		case 'x':
			_setbits (plc.flags, PLC_BAILOUT);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc != 1)
	{
		if (plc.nvm.file != -1)
		{
			error (1, ECANCELED, PLC_NODEVICE);
		}
	}
	openchannel (&channel);
	if (!(plc.message = malloc (sizeof (* plc.message))))
	{
		error (1, errno, PLC_NOMEMORY);
	}
	if (!argc)
	{
		Manager (&plc, ModuleID, ModuleSubID);
	}
	while ((argc) && (* argv))
	{
		if (!hexencode (channel.peer, sizeof (channel.peer), synonym (* argv, devices, SIZEOF (devices))))
		{
			error (1, errno, PLC_BAD_MAC, * argv);
		}
		Manager (&plc, ModuleID, ModuleSubID);
		argc--;
		argv++;
	}
	free (plc.message);
	closechannel (&channel);
	exit (0);
}

